### PR TITLE
api: improve docs for Trillian Log gRPC API

### PR DESCRIPTION
**Pull request by daviddrysdale**
- [ ] I have updated the [CHANGELOG](CHANGELOG.md).
  - Adjust the draft version number according to [semantic versioning](https://semver.org/) rules.
- [x] I have updated [documentation](docs/) accordingly.

### CODE REVIEW

**Comment by daviddrysdale** [comment]
Turns out server skew error behaviour wasn't quite as undocumented as I thought, on account of 69962cc7e4ed776ec5532a8381a44bcb66506add.

File: docs/api.md [hunk 1]
```diff
@@ -553,12 +567,13 @@ Continuing the CT example, for a CT mirror personality (which must allow dupes s
...
 ### Proof
-A consistency or inclusion proof for a Merkle tree. Output only.
+Proof holds a consistency or inclusion proof for a Merkle tree, as returned
+by the API.
 
 
 | Field | Type | Label | Description |
 | ----- | ---- | ----- | ----------- |
-| leaf_index | [int64](#int64) |  |  |
+| leaf_index | [int64](#int64) |  | leaf_index indicates the requested leaf index for leaf inclusion proof. This field is set to zero on a consistency proof. |
```
Comment (AlCutter) [inline docs/api.md]: preexisting: leaf_index == 0 is an entirely valid inclusion proof request and response, isn't it?  Context should probably be enough, but the wording here sounds a bit like a discriminator.

File: docs/api.md [hunk 2]
```diff
@@ -591,7 +606,7 @@ A consistency or inclusion proof for a Merkle tree. Output only.
 
 | Field | Type | Label | Description |
 | ----- | ---- | ----- | ----------- |
-| queued_leaf | [QueuedLogLeaf](#trillian.QueuedLogLeaf) |  |  |
+| queued_leaf | [QueuedLogLeaf](#trillian.QueuedLogLeaf) |  | queued_leaf describes the leaf is or will be incorporated into the Log. If the submitted leaf was already present in the Log (as indicated by its leaf identity hash), then the returned leaf will be the pre-existing leaf entry rather than the submitted leaf. |
```
Comment (AlCutter) [inline docs/api.md]: /the leaf is/the leaf which/

File: docs/api.md [hunk 3]
```diff
@@ -656,29 +671,64 @@ TODO(pavelkalinnikov): Consider renaming it to AddLogLeafResult or the like.
...
+The API supports adding new entries to the Merkle tree for a specific Log
+instance (identified by its log_id) in two modes:
+ - For a normal log, new leaf entries are queued up for subsequent
+   inclusion in the log, and the leaves are assigned consecutive leaf_index
+   values as part of that integration process.
+ - For a &#39;pre-ordered log&#39;, new entries have an already-defined leaf
+   ordering, and leaves are only integrated into the Merkle tree when a
+   contiguous range of leaves is available.
+
+The API also supports read operations to retrieve leaf contents, and to
```
Comment (AlCutter) [inline docs/api.md]: proofs of inclusion and the append-only...?

File: docs/api.md [hunk 4]
```diff
@@ -656,29 +671,64 @@ TODO(pavelkalinnikov): Consider renaming it to AddLogLeafResult or the like.
...
-| GetLeavesByHash | [GetLeavesByHashRequest](#trillian.GetLeavesByHashRequest) | [GetLeavesByHashResponse](#trillian.GetLeavesByHashResponse) | Returns a batch of leaves by their `merkle_leaf_hash` values. |
+| QueueLeaf | [QueueLeafRequest](#trillian.QueueLeafRequest) | [QueueLeafResponse](#trillian.QueueLeafResponse) | QueueLeaf adds a single leaf to the queue of pending leaves for a normal log. |
+| AddSequencedLeaf | [AddSequencedLeafRequest](#trillian.AddSequencedLeafRequest) | [AddSequencedLeafResponse](#trillian.AddSequencedLeafResponse) | AddSequencedLeaf adds a single leaf with an assigned sequence number to a pre-ordered log. |
+| GetInclusionProof | [GetInclusionProofRequest](#trillian.GetInclusionProofRequest) | [GetInclusionProofResponse](#trillian.GetInclusionProofResponse) | GetInclusionProof returns an inclusion proof for a leaf with a given index in a particular tree.
+
+If the requested tree_size is larger than the server is aware of, the response will include the known log root and an empty proof. |
```
Comment (AlCutter) [inline docs/api.md]: the latest known

File: docs/api.md [hunk 5]
```diff
@@ -656,29 +671,64 @@ TODO(pavelkalinnikov): Consider renaming it to AddLogLeafResult or the like.
...
+| GetInclusionProofByHash | [GetInclusionProofByHashRequest](#trillian.GetInclusionProofByHashRequest) | [GetInclusionProofByHashResponse](#trillian.GetInclusionProofByHashResponse) | GetInclusionProofByHash returns an inclusion proof for any leaves that have the given Merkle hash in a particular tree.
+
+If any of the leaves that match the given Merkle has have a leaf index that is beyond the requested tree size, the corresponding proof entry will be empty. |
+| GetConsistencyProof | [GetConsistencyProofRequest](#trillian.GetConsistencyProofRequest) | [GetConsistencyProofResponse](#trillian.GetConsistencyProofResponse) | GetConsistencyProof returns a consistency proof between different sizes of a particular tree.
+
+If the requested tree size is larger than the server is aware of, the response will include the known log root and an empty proof. |
```
Comment (AlCutter) [inline docs/api.md]: latest known log root

File: docs/api.md [hunk 6]
```diff
@@ -354,7 +354,7 @@ As an example, a Certificate Transparency frontend might set the following user
 | Field | Type | Label | Description |
 | ----- | ---- | ----- | ----------- |
 | signed_log_root | [SignedLogRoot](#trillian.SignedLogRoot) |  |  |
-| proof | [Proof](#trillian.Proof) |  | proof is filled if first_tree_size in GetLatestSignedLogRootRequest is non-zero. |
+| proof | [Proof](#trillian.Proof) |  | proof is filled if first_tree_size in GetLatestSignedLogRootRequest is non-zero (and within the tree size available at the server). |
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply docs/api.md]: Done.

File: docs/api.md (hunk 1 above)
Comment (daviddrysdale) [reply docs/api.md]: Reworded.

File: docs/api.md (hunk 3 above)
Comment (daviddrysdale) [reply docs/api.md]: Done.

File: docs/api.md (hunk 4 above)
Comment (daviddrysdale) [reply docs/api.md]: Done.

File: docs/api.md (hunk 5 above)
Comment (daviddrysdale) [reply docs/api.md]: Done.

**Comment by AlCutter** [comment]
> BTW, comments would be easier to apply if they were attached to the master source .proto file -- the generated api.md file seems to re-order things.

Yeah, soz - I just wrote comments on the files in the order they appeared, I realised once I'd reached the end that I should've written them on the proto instead :/

**Pull request merged**